File: rust/src/parse.rs
use std::iter;
use std::io::{Read, Seek, SeekFrom};
use std::str::FromStr;
use crate::app_error::AppError;

use crate::types::{raw, DiskVertex, Disk, LineItem, Model, ModelConfig};

pub fn parse_disks(
    items: &[LineItem],
    start: usize
) -> Option<(Vec<Disk>, usize)> {

    fn parse_vert(
        items: &[LineItem],
        start: usize
    ) -> Option<(DiskVertex, usize)> {
        let mut idx = start;

        let x = *if let Some(LineItem::Float(f)) = items.get(idx) {
            f
        } else {
            return None;
        };

        idx+= 1;
        let y = *if let Some(LineItem::Float(f)) = items.get(idx) {
            f
        } else {
            return None;
        };

        idx+= 1;
        let z = *if let Some(LineItem::Float(f)) = items.get(idx) {
            f
        } else {
            return None;
        };

        idx+= 1;
        let int = *if let Some(LineItem::Int(i)) = items.get(idx) {
            i
        } else {
            return None;
        };

        idx+= 1;
        Some((([x, y, z].into(), int), idx))
    }

    fn parse_disk(
        items: &[LineItem],
        expected_size: Option<usize>,
        start: usize,
    ) -> Option<(Disk, usize)> {
        let mut idx = start;

        let v_ct = *if let Some(LineItem::Int(i)) = items.get(idx) {
            i
        } else {
            return None;
        };

        if v_ct < 1 {
            return None;
        } 

        let v_ct = v_ct as usize;
        idx+= 1;

        if expected_size.is_some() && Some(v_ct) != expected_size {
            return None;
        }

        let mut disk = Disk::with_capacity(v_ct);
        
        for _ in 0..v_ct {
            if let Some((vert, new_idx)) = parse_vert(items, idx) {
                idx = new_idx;
                disk.push(vert);
            } else {
                return None;
            }
        }

        Some((disk, idx))
    }

    let mut idx = start;
    let mut disks = Vec::new();
    let mut disks_broken = true;

    while idx < items.len() {
        let disk_ct = if let Some(&LineItem::Int(count)) = items.get(idx) {
            count
        } else {
            0
        };

        idx+= 1;

        if disk_ct < 1 {
            continue;
        }

        let disk_ct = disk_ct as usize;
        let mut expected_count = None;
        disks_broken = false;

        for _ in 0..disk_ct {
            if let Some((disk, new_idx)) = parse_disk(
                items,
                expected_count,
                idx,
            ) {
                expected_count = Some(disk.len());
                disks.push(disk);
                idx = new_idx;
            } else {
                disks_broken = true;
                break;
            }
        }

        if !disks_broken {
            break;
        }
    }

    if disks_broken {
        None
    } else {
        Some((disks, idx))
    }
}

pub fn parse_body(
    items: &[LineItem],
    start: usize
) -> Option<(Vec<raw::BodySegment>, usize)> {
    let mut idx = start;

    let segment_ct = if let Some(&LineItem::Int(mut count)) = items.get(idx) {
        idx+= 1;

        if count < 0 {
            if let Some(&LineItem::Float(_)) = items.get(idx) {
                idx+= 1;
            } else {
                return None
            }

            if let Some(&LineItem::Int(ct)) = items.get(idx) {
                idx+= 1;
                count = ct;
            } else {
                return None
            }
        }
        
        count
    } else {
        return None;
    };

    let segment_ct = if let Ok(ct) = usize::try_from(segment_ct) {
        ct
    } else {
        return None;
    };

    let mut body = Vec::with_capacity(segment_ct);

    for _ in 0..segment_ct {
        let disk_info_idx = if let Some(&LineItem::Int(i)) = items.get(idx) {
            idx+= 1;
            i
        } else {
            return None;
        };

        let action = if let Some(&LineItem::Int(i)) = items.get(idx) {
            idx+= 1;
            i
        } else {
            return None;
        };

        let value = if let Some(&LineItem::Float(f)) = items.get(idx) {
            idx+= 1;
            f
        } else {
            return None;
        };

        let color = if let Some(&LineItem::Int(i)) = items.get(idx) {
            idx+= 1;
            i
        } else {
            return None;
        };

        let left = if let Some(&LineItem::Int(i)) = items.get(idx) {
            idx+= 1;
            i
        } else {
            return None;
        };

        let right = if let Some(&LineItem::Int(i)) = items.get(idx) {
            idx+= 1;
            i
        } else {
            return None;
        };

        let segment = raw::BodySegment {
            disk_info_idx,
            action,
            value,
            color,
            left,
            right
        };

        body.push(segment);
    }

    Some((body, idx))
}

pub fn parse_disk_info(
    items: &[LineItem],
    start: usize,
) -> Option<Vec<raw::DiskInformation>> {
    let mut idx = start;
    
    let info_ct = if let Some(&LineItem::Int(ct)) = items.get(idx) {
        idx+= 1;
        ct
    } else {
        return None;
    };

    let mut info = Vec::new();

    for _ in 0..info_ct {
        let mut arr1 = [0f32; 4];
        let mut arr2 = [0f32; 4];

        let shift_x = if let Some(&LineItem::Float(f)) = items.get(idx) {
            idx+= 1;
            f
        } else {
            return None;
        };

        let shift_y = if let Some(&LineItem::Float(f)) = items.get(idx) {
            idx+= 1;
            f
        } else {
            return None;
        };

        let scale_x = if let Some(&LineItem::Float(f)) = items.get(idx) {
            idx+= 1;
            f
        } else {
            return None;
        };

        let scale_y = if let Some(&LineItem::Float(f)) = items.get(idx) {
            idx+= 1;
            f
        } else {
            return None;
        };

        let disk_idx = if let Some(&LineItem::Int(i)) = items.get(idx) {
            idx+= 1;
            i
        } else {
            return None;
        };

        let id = if let Some(&LineItem::Int(i)) = items.get(idx) {
            idx+= 1;
            i
        } else {
            return None;
        };

        let flags = if let Some(&LineItem::Int(i)) = items.get(idx) {
            idx+= 1;
            i
        } else {
            return None;
        };

        for arr_idx in 0..4 {
            arr1[arr_idx] = if let Some(&LineItem::Float(f)) = items.get(idx) {
                idx+= 1;
                f
            } else {
                return None;
            };

            arr2[arr_idx] = if let Some(&LineItem::Float(f)) = items.get(idx) {
                idx+= 1;
                f
            } else {
                return None;
            };
        }

        info.push(raw::DiskInformation {
            shift: [shift_x, shift_y].into(),
            scale: [scale_x, scale_y].into(),
            disk_idx,
            id,
            flags,
            arr1,
            arr2,
        });
    }

    Some(info)
}

pub fn parse_lines<T: Read + Seek>(
    mut reader: T
) -> impl Iterator<Item = LineItem> {
    const BUF_MAX: usize = 256;
    let mut buffer = [0u8; BUF_MAX];
    let mut buffer_sz = 0usize;
    let mut record_start = reader.stream_position().unwrap();
    
    iter::from_fn(move || {
        let mut byte_buf = [0u8; 1];

        loop {
            let read_result = reader.read(&mut byte_buf[..]);

            match read_result {
                Ok(0) => { break; },
                Err(_) => { break; },
                _ => (),
            }

            let byte = byte_buf[0];

            if buffer_sz >= BUF_MAX || !byte.is_ascii() {
                let mut binary = Vec::new();

                if reader.seek(SeekFrom::Start(record_start)).is_ok() {
                    let _ = reader.read_to_end(&mut binary);
                }

                return Some(LineItem::Binary(binary));
            } else if  byte == b'\r' || byte == b'\n' {
                if byte == b'\r' {
                    if let Ok(n) = reader.read(&mut byte_buf) && n > 0 
                        && byte_buf[0] != b'\n'
                    {
                        reader.seek(SeekFrom::Current(-1)).unwrap();
                    }
                    
                    let string = String::from_utf8_lossy(&buffer[..buffer_sz]);
                    let string = string.trim().to_string();
                    buffer_sz = 0;
                    record_start = reader.stream_position().unwrap();

                    return Some(
                        if string.is_empty() {
                            LineItem::Empty
                        } else if let Ok(i) = i32::from_str(&string) {
                            LineItem::Int(i)
                        } else if let Ok(f) = f32::from_str(&string) {
                            LineItem::Float(f)
                        } else {
                            LineItem::Tag(string)
                        }
                    );
                }
            } else {
                buffer[buffer_sz] = byte;
                buffer_sz+= 1;
            }
        } 

        None
    })
}

pub fn parse_model(reader: impl Read + Seek) -> Result<Model, AppError> {
    let parsed_lines: Vec<_> = parse_lines(reader).collect();
    let mut idx = parsed_lines.len() - 1;
    
    while idx > 0 {
        if let LineItem::Tag(_) = parsed_lines[idx-1] {
            break;
        }

        idx-= 1;
    }

    let (disks, idx) = if let Some((disks, body_idx)) = parse_disks(
        &parsed_lines,
        idx
    ) {
        (disks, body_idx)
    } else {
        return Err(AppError("Parse failure: disks".to_string()));
    };

    let (body, idx) = if let Some((body, disk_info_idx)) = parse_body(
        &parsed_lines,
        idx,
    ) {
        (body, disk_info_idx)
    } else {
        return Err(AppError("Parse failure: body".to_string()));
    };

    let info = if let Some(info) = parse_disk_info(
        &parsed_lines,
        idx,
    ) {
        info
    } else {
        return Err(AppError("Parse failure: disk information".to_string()));
    };

    let config = ModelConfig {
        disks,
        disk_info: info,
        body
    };

    Ok(Model::new(config))
}

File: web/src/parse.ts
import {
    type LineItem,
    type Disk,
    type DiskVertex,
    type RawDiskInfo,
    type RawBodySegment,
    Model
} from "./types.ts";

import { Vec3 } from "./math.ts";

const ASCII = {
    CARRIAGE_RETURN: "\r".charCodeAt(0),
    LINE_FEED: "\n".charCodeAt(0),
    MAX: 127,
};

Object.freeze(ASCII);

export function parseModel(bytes: Uint8Array): Model {
    const parsedLines = parseLines(bytes);
    let idx = parsedLines.length - 1;

    while (idx > 0) {
        if (parsedLines[idx-1]?.kind === "tag") {
            break;
        }

        --idx;
    }

    // Skip a few records to prevent parsing non-disk records as disks
    idx+= 6;

    const [disks, bodyIdx] = parseDisks(parsedLines, idx);
    const [body, diskInfoIdx] = parseBody(parsedLines, bodyIdx);
    const diskInfo = parseDiskInfo(parsedLines, diskInfoIdx);

    return new Model({
        disks,
        diskInfo,
        body,
    });
}

export class ParseError extends Error {
    public line: number;

    constructor(reason: string, index: number) {
        const line = index + 1;
        super(`Line ${line}: ${reason}`);
        this.line = line;
    }
}

function parseDisks(lines: LineItem[], idx: number): [Disk[], number] {
    const errorMsg = "Failed to parse disks";

    const parseVert = (idx: number): [DiskVertex, number] => {
        let x: number, y: number, z: number, int: number;
        let item: LineItem | undefined;

        item = lines[idx];
        if (item?.kind === "float") {
            x = item.value;
            ++idx;
        } else {
            throw new ParseError(errorMsg, idx);
        }

        item = lines[idx];
        if (item?.kind === "float") {
            y = item.value;
            ++idx;
        } else {
            throw new ParseError(errorMsg, idx);
        }

        item = lines[idx];
        if (item?.kind === "float") {
            z = item.value;
            ++idx;
        } else {
            throw new ParseError(errorMsg, idx);
        }

        item = lines[idx];
        if (item?.kind === "int") {
            int = item.value;
            ++idx;
        } else {
            throw new ParseError(errorMsg, idx);
        }

        return [[new Vec3(x, y, z), int], idx];
    };

    const parseDisk = (
        idx: number,
        expectedSz: number | null
    ): [Disk, number] => {
        const disk = [];

        let vertCt: number;
        let item: LineItem | undefined;

        item = lines[idx];
        if (item?.kind === "int") {
            vertCt = item.value;
            ++idx;
        } else {
            throw new ParseError(errorMsg, idx);
        }

        if (vertCt < 1) {
            throw new ParseError(errorMsg, idx);
        }

        if (expectedSz != null && expectedSz !== vertCt) {
            throw new ParseError(errorMsg, idx);
        }

        for (let _i = 0; _i < vertCt; ++_i) {
            const [vert, newIdx] = parseVert(idx);
            idx = newIdx;
            disk.push(vert);
        }

        return [disk, idx];
    }

    let disks = [];
    let diskCt: number;
    let item: LineItem | undefined;
    let disksBroken = true;
    let expectedSz: number | null = null;

    while (idx < lines.length) {
        item = lines[idx];
        if (item?.kind === "int") {
            diskCt = item.value;
        } else {
            diskCt = 0;
        }

        ++idx;

        if (diskCt < 1) {
            continue;
        }

        let expectedCount: number | null = null;
        disksBroken = false;

        for (let _i = 0; _i < diskCt; ++_i) {
            try {
                const [disk, newIdx] = parseDisk(idx, expectedSz);
                expectedSz = disk.length;
                disks.push(disk);
                idx = newIdx;
            } catch (e) {
                disksBroken = true;
                break;
            }
        }

        if (disksBroken) {
            disks = [];
            expectedSz = null;
        } else {
            break;
        }
    }

    if (disksBroken) {
        throw new ParseError(errorMsg, idx);
    } else {
        return [disks, idx];
    }
}

function parseBody(lines: LineItem[], idx: number): [RawBodySegment[], number] {
    const body = [];
    const errorMsg = "Failed to parse body";

    let segmentCt: number;
    let item: LineItem | undefined;
    let diskInfoIdx: number, action: number, value: number, color: number,
        left: number, right: number;

    item = lines[idx];
    if (item?.kind === "int") {
        segmentCt = item.value;
        ++idx;

        while (segmentCt < 0) {
            item = lines[idx];
            if (item?.kind === "float") {
                ++idx;
            } else {
                throw new ParseError(errorMsg, idx);
            }

            item = lines[idx];
            if (item?.kind === "int") {
                segmentCt = item.value;
                ++idx;
            } else {
                throw new ParseError(errorMsg, idx);
            }
        }
    } else {
        throw new ParseError(errorMsg, idx);
    }

    for (let _i = 0; _i < segmentCt; ++_i) {
        item = lines[idx];
        if (item?.kind === "int") {
            diskInfoIdx = item.value;
            ++idx;
        } else {
            throw new ParseError(errorMsg, idx);
        }

        item = lines[idx];
        if (item?.kind === "int") {
            action = item.value;
            ++idx;
        } else {
            throw new ParseError(errorMsg, idx);
        }

        item = lines[idx];
        if (item?.kind === "float") {
            value = item.value;
            ++idx;
        } else {
            throw new ParseError(errorMsg, idx);
        }

        item = lines[idx];
        if (item?.kind === "int") {
            color = item.value;
            ++idx;
        } else {
            throw new ParseError(errorMsg, idx);
        }

        item = lines[idx];
        if (item?.kind === "int") {
            left = item.value;
            ++idx;
        } else {
            throw new ParseError(errorMsg, idx);
        }

        item = lines[idx];
        if (item?.kind === "int") {
            right = item.value;
            ++idx;
        } else {
            throw new ParseError(errorMsg, idx);
        }

        body.push({
            diskInfoIdx,
            action,
            value,
            color,
            left,
            right,
        });
    }

    return [body, idx];
}

function parseDiskInfo(lines: LineItem[], idx: number): RawDiskInfo[] {
    const diskInfo = [];
    const errorMsg = "Failed to parse disk info";

    let diskIdx: number, id: number, flags: number, infoCt: number;
    let item: LineItem | undefined;

    item = lines[idx];
    if (item?.kind === "int") {
        infoCt = item.value;
        ++idx;
    } else {
        throw new ParseError(errorMsg, idx);
    }

    for (let _i = 0; _i < infoCt; ++_i) {
        const shift: [number, number] = [0, 0];
        const scale: [number, number] = [0, 0];
        const arr1: [number, number, number, number] = [0, 0, 0, 0];
        const arr2: [number, number, number, number] = [0, 0, 0, 0];

        item = lines[idx];
        if (item?.kind === "float") {
            shift[0] = item.value;
            ++idx;
        } else {
            throw new ParseError(errorMsg, idx);
        }

        item = lines[idx];
        if (item?.kind === "float") {
            shift[1] = item.value;
            ++idx;
        } else {
            throw new ParseError(errorMsg, idx);
        }

        item = lines[idx];
        if (item?.kind === "float") {
            scale[0] = item.value;
            ++idx;
        } else {
            throw new ParseError(errorMsg, idx);
        }

        item = lines[idx];
        if (item?.kind === "float") {
            scale[1] = item.value;
            ++idx;
        } else {
            throw new ParseError(errorMsg, idx);
        }

        item = lines[idx];
        if (item?.kind === "int") {
            diskIdx = item.value;
            ++idx;
        } else {
            throw new ParseError(errorMsg, idx);
        }

        item = lines[idx];
        if (item?.kind === "int") {
            id = item.value;
            ++idx;
        } else {
            throw new ParseError(errorMsg, idx);
        }

        item = lines[idx];
        if (item?.kind === "int") {
            flags = item.value;
            ++idx;
        } else {
            throw new ParseError(errorMsg, idx);
        }

        const indices: [0, 1, 2, 3] = [0, 1, 2, 3];

        for (const arrIdx of indices) {
            item = lines[idx];
            if (item?.kind === "float") {
                arr1[arrIdx] = item.value;
                ++idx;
            } else {
                throw new ParseError(errorMsg, idx);
            }

            item = lines[idx];
            if (item?.kind === "float") {
                arr2[arrIdx] = item.value;
                ++idx;
            } else {
                throw new ParseError(errorMsg, idx);
            }
        }

        diskInfo.push({
            shift,
            scale: scale,
            diskIdx,
            id,
            flags,
            arr1,
            arr2,
        });
    }

    return diskInfo;
}

function parseLines(bytes: Uint8Array): LineItem[] {
    const maxBufSz = 256;
    const intRegex = /^-?\d+$/;
    const floatRegex = /^-?\d+\.\d+(e[-+]?\d+)?$/

    let items: LineItem[] = [];
    let lineBuffer = [];
    let itemStart = 0;
    let idx = 0;

    // Treat `bytes` like a stream, returning next byte on each call
    const read = (): number | null => {
        if (idx < bytes.length) {
            const byte = bytes[idx] ?? null;
            ++idx;
            return byte;
        }

        return null;
    };

    // Seek backwards by some number of elements
    const rewind = (count: number) => {
        idx-= count;

        if (idx < 0) {
            throw new Error("Rewound too far");
        }
    };

    // Read bytes into line items
    for (let byte; byte = read(), byte != null; ) {
        if (lineBuffer.length >= maxBufSz || byte > ASCII.MAX || byte === 0) {
            items.push({
                kind: "binary",
                value: bytes.subarray(itemStart)
            });

            // Not necessary, but it keeps state consistent
            idx = bytes.length;

            break;
        } else if (byte === ASCII.CARRIAGE_RETURN || byte === ASCII.LINE_FEED) {
            // Handle CR+LF (Windows) line-endings (which we expect from 3DA)
            if (byte === ASCII.CARRIAGE_RETURN) {
                byte = read();

                if (byte != null && byte !== ASCII.LINE_FEED) {
                    rewind(1);
                }
            }

            const str = charCodesToString(lineBuffer).trim();
            lineBuffer = [];
            itemStart = idx;

            if (str === "") {
                items.push({
                    kind: "empty",
                });
            } else if (intRegex.exec(str) != null) {
                items.push({
                    kind: "int",
                    value: parseInt(str, 10),
                });
            } else if (floatRegex.exec(str) != null) {
                items.push({
                    kind: "float",
                    value: parseFloat(str),
                })
            } else {
                items.push({
                    kind: "tag",
                    value: str,
                });
            }
        } else {
            lineBuffer.push(byte);
        }
    }

    return items;
}

const charCodesToString = (codes: number[]): string =>
    codes.map(code => String.fromCharCode(code)).join("");
